diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index e1a5e70a6fe..45b6f287b14 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -1484,10 +1484,18 @@ BOOL X11DRV_SetCursorPos( INT x, INT y )
         return FALSE;
     }

+#ifdef __ANDROID__
+    XNoOp( data->display );
+#else
     pXFixesHideCursor( data->display, root_window );
+#endif
     XWarpPointer( data->display, root_window, root_window, 0, 0, 0, 0, pos.x, pos.y );
     data->warp_serial = NextRequest( data->display );
+#ifdef __ANDROID__
+    XNoOp( data->display );
+#else
     pXFixesShowCursor( data->display, root_window );
+#endif
     XFlush( data->display ); /* avoids bad mouse lag in games that do their own mouse warping */
     TRACE( "warped to %d,%d serial %lu\n", x, y, data->warp_serial );
     return TRUE;
