diff --git a/dlls/nsiproxy.sys/ndis.c b/dlls/nsiproxy.sys/ndis.c
index 158a39e558d..b095467cbb7 100644
--- a/dlls/nsiproxy.sys/ndis.c
+++ b/dlls/nsiproxy.sys/ndis.c
@@ -23,7 +23,8 @@
 #endif

 #include "config.h"
-
+#include <ifaddrs.h>
+#include "nsi_common.h"
 #include <stdarg.h>
 #include <sys/types.h>
 #include <sys/socket.h>
@@ -104,7 +105,8 @@ struct if_entry

 static struct list if_list = LIST_INIT( if_list );
 static pthread_mutex_t if_list_lock = PTHREAD_MUTEX_INITIALIZER;
-
+static int do_not_update = -1;
+int wine_new_ndis = -1;
 static BOOL have_ethernet_iface;

 static struct if_entry *find_entry_from_index( UINT index )
@@ -309,6 +311,16 @@ static struct if_entry *add_entry( UINT index, const char *name )
         free( entry );
         return NULL;
     }
+#ifndef __ANDROID__
+    if_get_physical( name, &entry->if_type, &entry->if_phys_addr );
+#else
+    if (strstr(name, "wlan") || strstr(name, "en") || strstr(name, "eth") || strstr(name, "dummy"))
+        entry->if_type = MIB_IF_TYPE_ETHERNET;
+    else if (strstr(name, "lo"))
+        entry->if_type = MIB_IF_TYPE_LOOPBACK;
+    else if (strstr(name, "rmnet"))
+        entry->if_type = MIB_IF_TYPE_PPP;
+#endif

     if_get_physical( name, &entry->if_type, &entry->if_phys_addr );

@@ -330,9 +342,35 @@ static struct if_entry *add_entry( UINT index, const char *name )

 static unsigned int update_if_table( void )
 {
-    struct if_nameindex *indices = if_nameindex(), *entry;
+    struct if_nameindex *indices, *entry;
     unsigned int append_count = 0;
     struct if_entry *if_entry;
+    if (do_not_update == -1)
+        do_not_update = getenv("WINE_DO_NOT_UPDATE_IF_TABLE") && atoi(getenv("WINE_DO_NOT_UPDATE_IF_TABLE"));
+
+    if (wine_new_ndis == -1)
+        wine_new_ndis = getenv("WINE_NEW_NDIS") && atoi(getenv("WINE_NEW_NDIS"));
+
+    if (do_not_update)
+           return 0;
+
+    if (wine_new_ndis) {
+        struct ifaddrs *ifap;
+        int res;
+
+           TRACE( "Using alternative ndis implementation, output will be limited\n" );
+
+        if ((res = getifaddrs( &ifap )) == -1)
+            return 0;
+        while (ifap->ifa_next != NULL) {
+            int index = if_nametoindex( ifap->ifa_name );
+            if (index && !find_entry_from_index( index ) && add_entry( index, ifap->ifa_name ))
+                ++append_count;
+            ifap = ifap->ifa_next;
+        }
+        return append_count;
+    }
+    indices = if_nameindex();

     for (entry = indices; entry->if_index; entry++)
     {
@@ -532,7 +570,10 @@ static NTSTATUS ifinfo_enumerate_all( void *key_data, UINT key_size, void *rw_da
     pthread_mutex_lock( &if_list_lock );

     update_if_table();
-
+    if (do_not_update) {
+        pthread_mutex_unlock( &if_list_lock );
+        return STATUS_UNSUCCESSFUL;
+    }
     LIST_FOR_EACH_ENTRY( entry, &if_list, struct if_entry, entry )
     {
         if (num < *count)
