diff --git a/programs/winemenubuilder/winemenubuilder.c b/programs/winemenubuilder/winemenubuilder.c
index e874a903103..e514d2c6958 100644
--- a/programs/winemenubuilder/winemenubuilder.c
+++ b/programs/winemenubuilder/winemenubuilder.c
@@ -1082,8 +1082,7 @@ static HRESULT platform_write_icon(IStream *icoStream, ICONDIRENTRY *iconDirEntr
     LARGE_INTEGER zero;

     *nativeIdentifier = compute_native_identifier(exeIndex, icoPathW, destFilename);
-    iconsDir = heap_wprintf(L"%s", L"c:\\proton_shortcuts\\icons");
-    create_directories(iconsDir);
+    iconsDir = heap_wprintf(L"%s\\icons\\hicolor", xdg_data_dir);

     for (i = 0; i < numEntries; i++)
     {
@@ -1250,7 +1249,7 @@ static BOOL write_desktop_entry(const WCHAR *link, const WCHAR *location, const
     char *workdir_unix;
     int needs_chmod = FALSE;
     const WCHAR *name;
-    WCHAR *shortcuts_dir;
+    const WCHAR *prefix = _wgetenv( L"WINECONFIGDIR" );

     WINE_TRACE("(%s,%s,%s,%s,%s,%s,%s,%s,%s)\n", wine_dbgstr_w(link), wine_dbgstr_w(location),
                wine_dbgstr_w(linkname), wine_dbgstr_w(path), wine_dbgstr_w(args),
@@ -1258,12 +1257,11 @@ static BOOL write_desktop_entry(const WCHAR *link, const WCHAR *location, const
                wine_dbgstr_w(wmclass));

     name = PathFindFileNameW( linkname );
-
-    shortcuts_dir = heap_wprintf(L"%s", L"c:\\proton_shortcuts");
-    create_directories(shortcuts_dir);
-    location = heap_wprintf(L"%s\\%s.desktop", shortcuts_dir, name);
-    heap_free(shortcuts_dir);
-    needs_chmod = TRUE;
+    if(!location)
+    {
+        location = heap_wprintf(L"%s\\%s.desktop", xdg_desktop_dir, name);
+        needs_chmod = TRUE;
+    }

     file = _wfopen( location, L"wb" );
     if (file == NULL)
@@ -1272,9 +1270,15 @@ static BOOL write_desktop_entry(const WCHAR *link, const WCHAR *location, const
     fprintf(file, "[Desktop Entry]\n");
     fprintf(file, "Name=%s\n", wchars_to_utf8_chars(name));
     fprintf(file, "Exec=" );
+    if (prefix)
+    {
+        char *path = wine_get_unix_file_name( prefix );
+        fprintf(file, "env WINEPREFIX=\"%s\" ", path);
+        heap_free( path );
+    }

-    fprintf(file, "\"%s\"", escape(path));
-    if (args) fprintf(file, " \"%s\"", escape(args) );
+    fprintf(file, "wine %s", escape(path));
+    if (args) fprintf(file, " %s", escape(args) );
     fputc( '\n', file );
     fprintf(file, "Type=Application\n");
     fprintf(file, "StartupNotify=true\n");
